pkg/power/redfish/util.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package redfish
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+func getRedfishModel(access RedfishAccessInfo, endpoint string, model interface{}) error {
+	username := access.Username
+	password := access.Password
+	host := access.Host
+
+	// Create a HTTP client and set up the basic authentication header
+	client := &http.Client{}
+	url := host + endpoint
+	req, err := http.NewRequest("GET", url, http.NoBody)
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+	req = req.WithContext(ctx)
+
+	req.SetBasicAuth(username, password)
+
+	// Send the request and check the response
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Check the response status code
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("server returned status: %v", resp.Status)
+	}
+
+	// Decode the response body into the provided model struct
+	err = json.NewDecoder(resp.Body).Decode(model)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getRedfishSystem(access RedfishAccessInfo) (*RedfishSystemModel, error) {
+	var system RedfishSystemModel
+	err := getRedfishModel(access, "/redfish/v1/Systems", &system)
+	if err != nil {
+		return nil, err
+	}
+
+	return &system, nil
+}
+
+func getRedfishPower(access RedfishAccessInfo, system string) (*RedfishPowerModel, error) {
+	var power RedfishPowerModel
+	err := getRedfishModel(access, "/redfish/v1/Chassis/"+system+"/Power/PowerControl", &power)
```

Comment:
`/Power/PowerControl` is not specified as a resource in [Redfish Resource and Schema Guide](https://www.dmtf.org/sites/default/files/standards/documents/DSP2046_2023.1.pdf) and HPE iLo Redfish API does not provide it. 
cf. https://hewlettpackard.github.io/ilo-rest-api-docs/ilo5/#resource-map
Instead, all of `/Power/#PowerControl`, `/Power#PowerControl` and `/Power#/PowerControl` are available on my machine.  You can test HPE environment at https://ilorestfulapiexplorer.ext.hpe.com/ .

Reply:
thank you @tiwatsuka for the info! I am going to update the API model to work with your suggestion.

Reply:
@tiwatsuka can you get the output of `Power#/PowerControl` on your HPE? The simulator doesn't have it, thanks

Reply:
I can't access on my machine right now, but it was just same as output of `Power`. On second thoughts, it is just interpreted as a URL fragment that the part after `#`.

Reply:
```
{
  "@odata.context": "/redfish/v1/$metadata#Power.Power",
  "@odata.etag": "W/\"2613CD55\"",
  "@odata.id": "/redfish/v1/Chassis/1/Power",
  "@odata.type": "#Power.v1_3_0.Power",
  "Id": "Power",
  "Name": "PowerMetrics",
  "Oem": {
    "Hpe": {
      "@odata.context": "/redfish/v1/$metadata#HpePowerMetricsExt.HpePowerMetricsExt",
      "@odata.type": "#HpePowerMetricsExt.v2_3_0.HpePowerMetricsExt",
      "BrownoutRecoveryEnabled": true,
      "HasCpuPowerMetering": false,
      "HasDimmPowerMetering": false,
      "HasGpuPowerMetering": false,
      "HasPowerMetering": true,
      "HighEfficiencyMode": "Balanced",
      "Links": {
        "PowerMeter": {
          "@odata.id": "/redfish/v1/Chassis/1/Power/PowerMeter"
        },
        "FastPowerMeter": {
          "@odata.id": "/redfish/v1/Chassis/1/Power/FastPowerMeter"
        },
        "SlowPowerMeter": {
          "@odata.id": "/redfish/v1/Chassis/1/Power/SlowPowerMeter"
        },
        "FederatedGroupCapping": {
          "@odata.id": "/redfish/v1/Chassis/1/Power/FederatedGroupCapping"
        }
      },
      "MinimumSafelyAchievableCap": null,
      "MinimumSafelyAchievableCapValid": false,
      "SNMPPowerThresholdAlert": {
        "DurationInMin": 0,
        "ThresholdWatts": 0,
        "Trigger": "Disabled"
      }
    }
  },
  "PowerControl": [
    {
      "@odata.id": "/redfish/v1/Chassis/1/Power#PowerControl/0",
      "MemberId": "0",
      "PowerCapacityWatts": 500,
      "PowerConsumedWatts": 44,
      "PowerMetrics": {
        "AverageConsumedWatts": 40,
        "IntervalInMin": 20,
        "MaxConsumedWatts": 78,
        "MinConsumedWatts": 38
      }
    }
  ],
  "PowerSupplies": [
    {
      "@odata.id": "/redfish/v1/Chassis/1/Power#PowerSupplies/0",
      "FirmwareVersion": "1.00",
      "LastPowerOutputWatts": 44,
      "LineInputVoltage": 105,
      "LineInputVoltageType": "ACLowLine",
      "Manufacturer": "LTEON",
      "MemberId": "0",
      "Model": "865408-B21",
      "Name": "HpeServerPowerSupply",
      "Oem": {
        "Hpe": {
          "@odata.context": "/redfish/v1/$metadata#HpeServerPowerSupply.HpeServerPowerSupply",
          "@odata.type": "#HpeServerPowerSupply.v2_0_0.HpeServerPowerSupply",
          "AveragePowerOutputWatts": 44,
          "BayNumber": 1,
          "HotplugCapable": true,
          "MaxPowerOutputWatts": 58,
          "Mismatched": false,
          "PowerSupplyStatus": {
            "State": "Ok"
          },
          "iPDUCapable": false
        }
      },
      "PowerCapacityWatts": 500,
      "PowerSupplyType": "AC",
      "SerialNumber": "5WBXK0FLLG25CW",
      "SparePartNumber": "866729-001",
      "Status": {
        "Health": "OK",
        "State": "Enabled"
      }
    },
    {
      "@odata.id": "/redfish/v1/Chassis/1/Power#PowerSupplies/1",
      "MemberId": "1",
      "Oem": {
        "Hpe": {
          "@odata.context": "/redfish/v1/$metadata#HpeServerPowerSupply.HpeServerPowerSupply",
          "@odata.type": "#HpeServerPowerSupply.v2_0_0.HpeServerPowerSupply",
          "BayNumber": 2
        }
      },
      "Status": {
        "Health": "Warning",
        "State": "Absent"
      }
    }
  ]
}
```

Reply:
thank you @tiwatsuka this looks consistent with the new model definition [here](https://github.com/sustainable-computing-io/kepler/pull/734/files#diff-1c7ba32c2f6c4c86fc187fd007b4ce14aa33f9e481377e7a2b111ee86a520db3R74)

I'll finish it next week and hopefully get into the next release in mid July.